File: 11-构建LaTeX/main.py
import os

fileignore = [
    ".git", ".font", "01-STL", "11-构建LaTeX", "in", "00-封装类",
    "in.txt", "out.txt", ".ogmc", "o-o", ".gitignore"
]

# 向要生成的模板写入准备好的头部与尾部
def writetex(filename):
    fp = open(filename, "r")
    while True:
        buff = fp.readline()
        if len(buff) == 0:
            break
        tex.write(buff)
    fp.close()


def getLaTeX(path, depth):
    files = os.listdir(path)                        # 得到文件夹下的所有文件(包括文件夹)名称
    files.sort()                                    # 排序
    for file in files:                              # 遍历当前文件夹下所有文件
        if os.path.isdir(path + file):              # 如果是文件夹，递归
            if file in fileignore:                  # 排除无关文件夹
                continue
            if depth == 0:                          # 一级目录特判
                tex.write("\\section{" + file + "}\n")
            else:
                str = "sub"
                for i in range(1, depth):
                    str += "sub"
                if (len(str) > 6):                  # 三级目录特判
                    tex.write("\\paragraph{" + file + "}\n")
                    continue
                tex.write("\\" + str + "section{" + file + "}\n")
            getLaTeX(path + file + "/", depth + 1)  # 递归寻找下一级目录
        elif os.path.isfile(path + file):           # 如果是文件
            if file in fileignore:                  # 排除无关文件夹
                continue
            #  if depth == 1:
            #      str = "sub"
            #  elif depth == 2:
            #      str = "subsub"
            str = "sub"
            for i in range(1, depth):
                str += "sub"
            if (len(str) > 6):
                tex.write("\\paragraph{" + file + "}\n")
                tex.write("\\lstinputlisting{" + "." + path + file + "}\n")
                continue
            tex.write("\\" + str + "section{" + file + "}\n")
            tex.write("\\lstinputlisting{" + "." + path + file + "}\n")


os.chdir("..")                                      # 设置路径为..
tex = open("./11-构建LaTeX/template.tex", "w")      # 打开template.tex文件
writetex("./11-构建LaTeX/LaTeX-Head")               # 写入准备好的模板头
path = "./"                                         # 路径还是..
getLaTeX(path, 0)                                   # 从path开始写入模板
writetex("./11-构建LaTeX/LaTeX-Tail")               # 写入准备好的模板尾
tex.close()                                         # 关闭文件

File: 11-构建LaTeX/bin/main.cpp
#include <cstdio>
#include <cstring>
const int maxn = 1000;
int cnt;
char sec[100][maxn];
FILE* tex = NULL;
char buff[maxn];
char File_name[maxn], File_file[maxn];

void File_div(const char s[]) {
    int nl = 0, fl = 0;
    int i = 0;
    while (s[i] && s[i] != '<') ++i;
    if (s[i]) ++i;
    while (s[i] && s[i] != '>')
        File_name[nl++] = s[i++];
    while (s[i] && s[i] != '<') ++i;
    if (s[i]) ++i;
    while (s[i] && s[i] != '>')
        File_file[fl++] = s[i++];
    File_name[nl] = File_file[fl] = 0;
}

void readContents() {
    FILE* dirname = fopen("Contents", "r");
    while (fgets(sec[cnt++], maxn, (FILE*)dirname) != NULL) {
        continue;
    }
    fclose(dirname);
    //去除最后一个
    while (sec[cnt-1][0] == 0 && cnt > 0)
        --cnt;
    for (int i = 0; i < cnt; ++i)
        //如果最后一位不是回车，就换成回车
        if (sec[i][strlen(sec[i])-1] != 0)
            sec[i][strlen(sec[i])-1] = 0;
}

void LaTeXFile(const char* filename) {
    FILE* fp = fopen(filename, "r");
    while (fgets(buff, maxn, fp) != NULL)
        fprintf(tex, "%s", buff);
    fclose(fp);
}

int main() {
    tex = fopen("template.tex", "w");
    LaTeXFile("LaTeX-Head");
    readContents();
    for (int i = 0; i < cnt; ++ i) {
        fprintf(tex, "\\section{%s}\n", sec[i]);
        memcpy(buff, sec[i], sizeof buff);
        strcat(buff, "/Files");
        FILE* fp = fopen(buff, "r");
        while(fgets(buff, maxn, fp)) {
            if (buff[0] != '<') break;
            // lb.solve(buff);
            File_div(buff);
            fprintf(tex, "\\subsection{%s}\n", File_name);
            memcpy(buff, sec[i], sizeof buff);
            //添加回车，否则strcat就会把后面的也连接起来
            buff[strlen(buff)+1] = 0;
            buff[strlen(buff)] = '/';
            strcat(buff, File_file);
            fprintf(tex, "\\lstinputlisting{%s}\n", buff);
        }
        fclose(fp);
    }
    LaTeXFile("LaTeX-Tail");
    fclose(tex);
    return 0;
}
